### PR TITLE
MM-54558 SSO code challenge verification

### PR DESCRIPTION
**Pull request by enahum**
#### Summary
Adds code challenge verification for SSO logins using the web browser.

Applies to OAuth, OpenID and SAML

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-54558

Server PR: https://github.com/mattermost/mattermost/pull/24734

#### Release Note
```release-note
Added code challenge verification for SSO login
```

### CODE REVIEW

**Review by larkox — APPROVED** [review]

**Review by esarafianou — APPROVED** [review]
LGTM 🚀

File: app/screens/sso/sso_with_redirect_url.tsx [hunk 1]
```diff
@@ -62,23 +66,26 @@ const SSOWithRedirectURL = ({doSSOLogin, loginError, loginUrl, serverUrl, setLog
     const [error, setError] = useState<string>('');
     const style = getStyleSheet(theme);
     const intl = useIntl();
+    const codeVerifier = useMemo(() => generateRandomString(100), []);
```
Comment (jupenur) [inline app/screens/sso/sso_with_redirect_url.tsx:69]: The memoization here means the verifier will be only generated once and could be reused in multiple SSO flows, right? Including across different servers? That's not good.

File: app/utils/strings.ts [hunk 2]
```diff
@@ -4,3 +4,15 @@
...
     return s.replace(' ', '\xa0');
 }
+
+export function generateRandomString(length: number) {
+    let result = '';
+    let counter = 0;
+    const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+    const charsLenght = chars.length;
+    while (counter < length) {
+        result += chars.charAt(Math.floor(Math.random() * charsLenght));
```
Comment (jupenur) [inline app/utils/strings.ts:14]: This needs to use a CSPRNG. `Math.random` is notoriously predictable.

**Review by jupenur — CHANGES_REQUESTED** [review]

File: app/screens/sso/sso_with_redirect_url.tsx (hunk 1 above)
Comment (enahum) [reply app/screens/sso/sso_with_redirect_url.tsx:69]: No, is memoized only while the component is in memory, changing servers/ going back to the previous screen removes the verifier from memory

File: app/utils/strings.ts (hunk 2 above)
Comment (enahum) [reply app/utils/strings.ts:14]: Can you double check this lib? https://www.npmjs.com/package/react-native-securerandom

**Comment by enahum** [comment]
Closing as won't accept

**Pull request closed without merging**